contrib/hyperscan/matching/input_matchers/test/matcher_speed_test.cc

```diff
@@ -0,0 +1,57 @@
+// Note: this should be run with --compilation_mode=opt, and would benefit from
+// a quiescent system with disabled cstate power management.
+
+#include "source/common/common/assert.h"
+#include "source/common/common/regex.h"
+
+#include "benchmark/benchmark.h"
+#include "contrib/hyperscan/matching/input_matchers/source/matcher.h"
+
+namespace Envoy {
+
+static const std::vector<std::string> ClusterInputs = {
+    "cluster.no_trailing_dot",
+    "cluster.match.",
+    "cluster.match.normal",
+    "cluster.match.and.a.whole.lot.of.things.coming.after.the.matches.really.too.much.stuff",
+};
+
+static const std::string ClusterRePattern = "^cluster\\.((.*?)\\.)";
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+static void BM_CompiledGoogleReMatcher(benchmark::State& state) {
+  envoy::type::matcher::v3::RegexMatcher config;
+  config.mutable_google_re2();
+  config.set_regex(ClusterRePattern);
+  const auto matcher = Regex::CompiledGoogleReMatcher(config);
+  uint32_t passes = 0;
+  for (auto _ : state) { // NOLINT
+    for (const std::string& cluster_input : ClusterInputs) {
+      if (matcher.match(cluster_input)) {
+        ++passes;
+      }
+    }
+  }
+  RELEASE_ASSERT(passes > 0, "");
+}
+BENCHMARK(BM_CompiledGoogleReMatcher);
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+static void BM_HyperscanMatcher(benchmark::State& state) {
+  envoy::extensions::matching::input_matchers::hyperscan::v3alpha::Hyperscan config;
+  auto regex = config.add_regexes();
+  regex->set_regex(ClusterRePattern);
+  auto matcher = Extensions::Matching::InputMatchers::Hyperscan::Matcher(config);
+  uint32_t passes = 0;
+  for (auto _ : state) { // NOLINT
+    for (const std::string& cluster_input : ClusterInputs) {
+      if (matcher.match(cluster_input)) {
+        ++passes;
+      }
+    }
+  }
+  RELEASE_ASSERT(passes > 0, "");
+}
+BENCHMARK(BM_HyperscanMatcher);
```

Comment:
I suggest making these benchmarks multithreaded if you need synchronization during the match.

Reply:
Thanks for your suggestion, I updated the benchmark tests and the result shows their is little performance degradation.
```
----------------------------------------------------------------------------------------------
Benchmark                                                    Time             CPU   Iterations
----------------------------------------------------------------------------------------------
BM_CompiledGoogleReMatcher/process_time/threads:20        47.3 ns          691 ns      1388140
BM_CompiledGoogleReMatcher/process_time/threads:200       24.1 ns          369 ns      2000000
BM_HyperscanMatcher/process_time/threads:20               30.8 ns          450 ns      2423240
BM_HyperscanMatcher/process_time/threads:200              16.4 ns          251 ns      2825000
```

Reply:
Well, it is because each thread will have its own `matcher` object, and they will all have their own mutex. Benchmark engine will simply create N threads and call the benchmarking function in each thread. To see contention you need to make the `matcher` object `static` so it is hared across all benchmarking threads.

Reply:
Yes, I have not noticed that. I would like to switch to a lock-free way (like use a lock-free queue) to avoid the degradation.

Reply:
Do we have a existing implementation of a lock-free queue, or one that can be picked up directly from absl? If not, can I implement one or import one as a dependency?

Reply:
There is another approach to use the thread local storage, but we will not able to benchmark it multithreaded.

Reply:
Thread local storage is a good solution for this and Envoy already has an implementation for it. I would go with it. And if there is no more state shared across threads, then there should be no need for multithreaded testing, shouldn't it?

Reply:
Ok, I will use the thread local storage.

Reply:
I would actually keep the mulththreaded tests with a static instance on matcher, since this will produce the most relevant result.